gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcWriteChannel.java

```diff
@@ -178,8 +192,12 @@ private Object doResumableUpload() throws IOException {
       long writeOffset = 0;
       int retriesAttempted = 0;
       Hasher objectHasher = Hashing.crc32c().newHasher();
+      // Holds list of most recent number of numberOfRequestsToRetain requests, so upload can be
+      // rewound and re-sent upon non-transient errors.
+      TreeMap<Long, ByteString> dataChunkMap = new TreeMap<>();
```

Comment:
I assume the reason you chose a TreeMap is because you can then key on the offset. But since you always just iterate over it (rather than do a single lookup to find the needed chunk), I think this is overkill and kind of confusing; all you really need is a LinkedList, so you know you will iterate in offset order.

Reply:
But we need to keep both value of offset, not just order, along with cached data chunk. Just the fact that data chunks are pushed in increasing offset order will not be enough to pick the right data chunk to resume write.

Reply:
Oh, I see, your map no longer stores the pre-built request objects, but rather just the bytes (which is why you then also need to store the offset). In my implementation I held on to the request objects. The way you are doing it forces you to re-scan the data every time you re-transmit (to compute the checksums).